### PR TITLE
Snapshot contract between the config builder and the runtime.

### PR DESCRIPTION
**Pull request by guptasu**
Add new configuration data to the snapshot. Ephemeral.go builds the snapshot. Snapshot is used by the runtime dispatch table.

Following this PR, @mandarjog will work on consuming this snapshot during request path and I will work on building this snapshot as we load the config.

### CODE REVIEW

File: mixer/pkg/runtime/config/snapshot.go [hunk 1]
```diff
@@ -97,6 +144,49 @@ type (
...
+
+		SupportedTemplates []Template
+
+		SessionBased bool
+
+		Description string
+	}
+
+	// Client contains the connection data for the adapter
+	Client struct {
```
Comment (ozevren) [inline mixer/pkg/runtime/config/snapshot.go]: I don't think this is the right place have a live connection. The config stuff should be about configuration only, not runtime state. We probably need something similar to handler table forthis.

File: mixer/pkg/runtime/config/snapshot.go [hunk 2]
```diff
@@ -44,10 +47,27 @@ type (
 		//  Adapters2 contains adapter metadata loaded from the store
```
Comment (ozevren) [inline mixer/pkg/runtime/config/snapshot.go]: I am trying to understand why we have the things with the "2" suffix. Are these going to replace the non-suffixed ones in the future? If so, we should add defaults for those.

If some of these are going to be living side-by-side (i.e. Adapters & Templates), then we should use better naming.

Comment (guptasu) [reply mixer/pkg/runtime/config/snapshot.go]: In my opinion the 2 suffix will replace the old ones. @ozevren what do you mean by `add defaults for those`, not sure I follow ?

File: mixer/pkg/runtime/config/snapshot.go (hunk 1 above)
Comment (guptasu) [reply mixer/pkg/runtime/config/snapshot.go]: You are right.

File: mixer/pkg/runtime/config/snapshot.go [hunk 3]
```diff
@@ -61,6 +81,16 @@ type (
...
 	}
 
+	// Instance2 configuration. Fully resolved.
+	Instance2 struct {
+		Name string
+
+		Template *Template
+
+		// Encoder to create request instance bytes from attributes
+		Encoder *dynamic.Encoder
```
Comment (mandarjog) [inline mixer/pkg/runtime/config/snapshot.go]: `Encoder *dynamic.Encoder` is not part of the configuration.
instance2 should have name, template and input to the EncoderBuilder.
We should create the Encoder as part of creating the runtime object.

Comment (guptasu) [reply mixer/pkg/runtime/config/snapshot.go]: For the purpose of validating the user written instance, we will be creating the `Encoder` during snapshot building phase. IMO encoder is not super different that the user config input, it is just a different view of it, why should we not just cache it inside snapshot itself ? WDYT ?

Comment (mandarjog) [reply mixer/pkg/runtime/config/snapshot.go]: what is the equivalent field on the generated code side?

Comment (guptasu) [reply mixer/pkg/runtime/config/snapshot.go]: There is a builder that we compute during a routing table building phase..
https://github.com/istio/istio/blob/66434029ad81f1d6e049c413c5150908d147d9d1/mixer/pkg/runtime/routing/builder.go#L223

We should do this `builder` creation during snapshot building phase. During routing table building we should just connect to the handler, build it and reference the pre built artifacts inside the routing table entries. Done.

**Comment by mandarjog** [comment]
unrelated pilot e2e failure.

File: mixer/pkg/runtime/config/snapshot.go (hunk 3 above)
Comment (mandarjog) [reply mixer/pkg/runtime/config/snapshot.go]: He is storing the result of validation which is the encoder and an error.
Validation can throw away the result and call build again at a later time.

```
encoder, err := builder.build(protodecs, bytes)
```

**Pull request merged**